Comfyui serverless runpod image generation

9nbf7c4q6b-lgtm opened the pull request:
Adds support to allow calling a ComfyUI Serverless endpoint on RunPod for image generation.

# Use case
* Allows using the same ComfyUI workflows for the local ComfyUI setup but with remote resources.
  * Offload memory usage for image generation.
  * Run larger models than your local machine is capable of supporting.
  * Run models on faster hardware than you have locally.

# Setup
* You'll need a runpod account and to add some money to it. I'm seeing around 2 cents per image for Qwen image generation so $5 would last a while.
* https://console.runpod.io/hub/runpod-workers/worker-comfyui is a flux1 dev configuration that you can use to create your own serverless endpoint.
  * There is information there on creating your own configuration if you want to use a different model or add loras.
* Create an API key for access to the serverless endpoint. https://console.runpod.io/user/settings

* In ST, select ComfyUI RunPod as the source.
* Set the URL to the URL of your endpoint.
* Set the API key.
* Click Connect. If the API key and url are correct, you should get toasts indicating success.
* The ComfyUI workflow configuration flow is the same as local.
  * Use the "Export (API)" option.
  * Model, samplers, vae, etc cannot be determined so your workflow needs to have these hard coded (no %model% substitution).
  * Other substitutions should work the same as local.

# Extra notes
* The serverless configuration does not currently embed the workflow into the output image. Ie, you won't be able to drag/drop the image into local comfyui to see the seed or prompt. This is just a limitation of the runpod handler and is a capability that could be added on that side.

## Checklist:

- [x] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

9nbf7c4q6b-lgtm committed:
Add comfyrunpod generation endpoints.

Get a bunch of settings added.

Probably get runpod api secret stored properly.

sending to runpod.

Correct workflow wrapping for runpod.

9nbf7c4q6b-lgtm committed:
Reuse comfyui workflow setting block for runpod.

9nbf7c4q6b-lgtm committed:
Update API key url to point to runpod settings

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -1718,6 +1755,11 @@ async function loadModels() {
         case sources.comfy:
             models = await loadComfyModels();
             break;
+        case sources.comfy_runpod:
```
Something is wrong here.

9nbf7c4q6b-lgtm replied:
Separate from the duplicate case, I'm copying the pattern other services are using by having loadComfyRunPodModels update the "Click to Set" text if the API key is set.

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -2982,6 +3040,7 @@ async function sendGenerationRequest(generationType, prompt, additionalNegativeP
         }
 
         if (!result.data) {
+            console.log(JSON.stringify(result));
```
Remove the debug log?

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -3765,6 +3824,98 @@ async function generateComfyImage(prompt, negativePrompt, signal) {
+        const text = await workflowResponse.text();
+        toastr.error(`Failed to load workflow.\n\n${text}`);
+    }
+    let workflow = (await workflowResponse.json()).replaceAll('"%prompt%"', JSON.stringify(prompt));
```
Almost 1:1 duped code with regular comfy. Do they really need to be split? Can you maybe create a function to get a filled out workflow string?

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -4299,6 +4450,8 @@ function isValidState() {
         case sources.comfy:
             return true;
+        case sources.comfy_runpod:
+            return true;
```
I'd say that it can only be valid if a URL is provided Same for regular comfy.

Cohee1207 reviewed (changes requested):
Too much duped code. I'd say add a checkbox that enables serverless mode. You can't use them at the same time with regular anyway (since only one source can be active).

9nbf7c4q6b-lgtm committed:
Remove debug print.

9nbf7c4q6b-lgtm committed:
Combine normal and runpod comfyui generation functions

9nbf7c4q6b-lgtm committed:
Remove unused server endpoints for runpod.

9nbf7c4q6b-lgtm committed:
Remove debug prints from server.

9nbf7c4q6b-lgtm committed:
Remove redundant switch case.

9nbf7c4q6b-lgtm committed:
Check for url to see if image generation is possible for comfy and comfy runpod.

9nbf7c4q6b-lgtm committed:
lint

9nbf7c4q6b-lgtm commented:
@Cohee1207 Thanks for the review. 
* I addressed all of the comments to some degree.
* While the workflow input is the same between the two, I think most of what is remaining is different enough to justify a second service entry. The separate API key and network handling in particular I think are cleaner as separate handlers. I can look at making it a checkbox if you still think it is warranted.

9nbf7c4q6b-lgtm committed:
Correct/update doc comments.

Cohee1207 commented:
I think it will be much easier to review and maintain if it remains the same provider. Maybe a dropdown select e.g. mode, like: Local, Runpod that switches between different UI elements? You can keep the URLs different even and only show API key input in the second mode.

9nbf7c4q6b-lgtm committed:
Make runpod serverless a sub-type under comfyui source in the UI.

9nbf7c4q6b-lgtm commented on public/scripts/extensions.js:
```diff
@@ -1294,6 +1294,9 @@ export async function loadExtensionSettings(settings, versionChanged, enableAuto
     }
+    if (extension_settings.sd.source === 'comfy' && !extension_settings.sd.comfy_type) {
+        extension_settings.sd.comfy_type = 'standard';
+    }
```
This doesn't seem like how I should do this, but not sure where a better place would be. It's looking to make sure that the default 'standard' value is set.

Cohee1207 replied:
There is already existing code which sets default values. Add it to `const defaultSettings`

9nbf7c4q6b-lgtm commented:
@Cohee1207 Revised UI and settings to have serverless option be a sub-type of the ComfyUI source.

<img width="395" height="396" alt="image" src="https://github.com/user-attachments/assets/8b0a8755-d24d-449e-a6b6-c5d96906ed2b" />

<img width="366" height="439" alt="image" src="https://github.com/user-attachments/assets/301ab62e-d53f-481f-9639-2785cc91e031" />

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -320,6 +324,8 @@ const defaultSettings = {
     comfy_url: 'http://127.0.0.1:8188',
     comfy_workflow: 'Default_Comfy_Workflow.json',
 
+    comfy_runpod_url: 'https://api.runpod.ai/v2/iunj07kvan2sed',
```
Are you sure this should be here? looks like your personal URL.

9nbf7c4q6b-lgtm replied:
You need a token to access anyways so the URL doesn't really matter. Do you think something like:
`https://api.runpod.ai/v2/[your endpoint id]` would be clear?

Cohee1207 replied:
Leave it empty. Put example value into HTML placeholder.

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -2210,6 +2260,12 @@ function loadNovelSchedulers() {
 
 async function loadComfyModels() {
+    if (extension_settings.sd.comfy_type === comfyTypes.runpod_serverless) {
+        $('#sd_runpod_key').toggleClass('success', !!secret_state[SECRET_KEYS.RUNPOD]);
```
Is runpod key the same for all services? Does it make more sense to scope it to imag.gen extension specifically?

9nbf7c4q6b-lgtm replied:
This runpod key is specific to runpod. It would be different from say a Black Forest labs key.

Cohee1207 replied:
I understand. I mean is it only issued once by runpod or is it specific to comfy that runs on runpod?

9nbf7c4q6b-lgtm replied:
Runpod allows for multiple keys to be created with different access levels. Realistically, a person will probably reuse the same key to allow for access to multiple runpod endpoints. They could have an LLM serverless endpoint setup also and use the same API key.

Cohee1207 replied:
Let's have them scoped to comfy for now as there's nothing else with Runpod support in mind.

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -2456,7 +2518,7 @@ async function loadComfyVaes() {
 
 async function loadComfyWorkflows() {
-    if (!extension_settings.sd.comfy_url) {
+    if (!extension_settings.sd.comfy_url || !extension_settings.sd.comfy_runpod_url) {
```
Actually loading workflows is not limited to only the cases where the URL is provided since it's loaded from ST's local server. So this whole condition can be removed.

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -2947,7 +3009,16 @@ async function sendGenerationRequest(generationType, prompt, additionalNegativeP
+                        result = await generateComfyRunPodImage(prefixedPrompt, negativePrompt, signal);
+                        break;
+                    case comfyTypes.standard:
+                        // fallthrough
```
Maybe throw an error in fallthrough instead?

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -4298,7 +4401,15 @@ function isValidState() {
+                case comfyTypes.runpod_serverless:
+                    return !!extension_settings.sd.comfy_runpod_url &&
+                        secret_state[SECRET_KEYS.RUNPOD];
+                case comfyTypes.standard:
```
Same here. Just return false in fallthrough case.

Cohee1207 commented on public/scripts/secrets.js:
```diff
@@ -109,6 +110,7 @@ const FRIENDLY_NAMES = {
     [SECRET_KEYS.CUSTOM_OPENAI_TTS]: 'Custom OpenAI TTS',
     [SECRET_KEYS.TAVILY]: 'Tavily',
     [SECRET_KEYS.BFL]: 'Black Forest Labs',
+    [SECRET_KEYS.RUNPOD]: 'ComfyUI RunPod',
```
An extension of my previous comment: the text says ComfyUI RunPod, but the key is just called 'runpod'. Maybe name it COMFY_RUNPOD?

9nbf7c4q6b-lgtm committed:
Set comfy_type in stable-diffusion defaultSettings

9nbf7c4q6b-lgtm committed:
Allow loading comfy workflows even if url isnt set yet.

9nbf7c4q6b-lgtm committed:
Change default runpod url to template example

9nbf7c4q6b-lgtm committed:
Rename runpod api key to comfy_runpod

9nbf7c4q6b-lgtm committed:
Handle default switches separately from standard now that default setting is ensured.

9nbf7c4q6b-lgtm committed:
Update runpod api key name on server.

9nbf7c4q6b-lgtm committed:
Move example runpod url to placeholder.

9nbf7c4q6b-lgtm committed:
lint

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -1166,6 +1185,12 @@ function onComfyUrlInput() {
 }
 
+function onComfyRunPodUrlInput() {
+    // Remove trailing slashes
```
I don't think it does that actually?..

Cohee1207 reviewed (approved):
The code looks fine. I'll try to test it before merging.

9nbf7c4q6b-lgtm committed:
Remove incorrect comment.

Cohee1207 committed:
Merge branch 'staging' into comfyui-runpod-generation

Cohee1207 committed:
Hide toasts remote models toasts

Cohee1207 committed:
React to key updates

Cohee1207 committed:
Add type annotation for data and improve workflow handling in generate endpoint

Cohee1207 commented:
Yeah it took some time to figure out, but I got it set up and tested. You might want to put the part from PR intro post into the doc: https://github.com/SillyTavern/SillyTavern-Docs/blob/main/extensions/Stable-Diffusion.md
